apps/tx-builder/src/pages/Dashboard.tsx

```diff
@@ -62,17 +36,76 @@ const Dashboard = (): ReactElement => {
     setContract(interfaceRepo.getMethods(abi))
   }, [abi, interfaceRepo])
 
-  const isAddressInputFieldValid = isValidAddress(address) || !address
+  const isAbiAddressInputFieldValid = !abiAddress || isValidAddress(abiAddress)
 
-  const contractHasMethods = contract && contract.methods.length > 0 && !isABILoading
+  const contractHasMethods =
+    abiStatus === FETCH_STATUS.SUCCESS && contract && contract.methods.length > 0
 
-  const isTransferTransaction = isValidAddress(address) && !abi && !isABILoading
-  const isContractInteractionTransaction = abi && contract && !isABILoading
+  const isTransferTransaction =
+    abiStatus === FETCH_STATUS.SUCCESS && isAbiAddressInputFieldValid && !abi
+  const isContractInteractionTransaction = abiStatus === FETCH_STATUS.SUCCESS && abi && contract
 
   const showNewTransactionForm = isTransferTransaction || isContractInteractionTransaction
 
   const showNoPublicMethodsWarning = contract && contract.methods.length === 0
 
+  // The flow is the following:
+  // 1. We receive the address input, we check if it's a valid address
+  // 2. If it's a valid address, we check if it's a proxy
+  // 3. If it's a proxy, we get the implementation address and check if there's an ABI for it
+  // 4. If there's an ABI for the implementation address, we show the ABI dialog
+  const handleAddressInput = useCallback(
```

Comment:
As a suggestion you can keep the same logic that we had before and just create an `useEffect`that only detects if its a proxy and show the popup like this:

```
  const {
    open: showProxyModal,
    openModal: openProxyModal,
    closeModal: closeProxyModalModal,
  } = useModal()

  useEffect(() => {
    if (isValidAddress(address) && !!abi) {
      // we check if a proxy is present
      const proxyAddress = detectProxyTarget()

      if (proxyAddress) {
        openProxyModal()
      }
    }
  }, [abi, address])
```

Reply:
This is the first thing I tried to do, but it turned out the component into a complete mess of states and effects. One thing triggers one effect, that effect triggers another, and it becomes very hard to keep track of the current state. Also, I don't think the effect is the right way to go, let's compare two flows:
1. With an effect
```
1. We get an onChange event
2. We set the address in the state
3. We run an effect to potentially update the `address` state again
4. We update the `address` state again
```
2. Without an effect
```
1. We get an onChange event
2. We calculate new state based on the address in the event
3. We set new state
```

In my opinion, **2)** is more pragmatic and straightforward than having multiple effects triggering one another.

Reply:
Also, I told you that the `AddressInput` component triggers duplicate events, so if we have chained but separate effects it adds even more complexity to prevent them from running unnecessarily

Reply:
What is your opinion about to use a different state variable `implementationProxyAddress` to track the implementation address like I suggested in [this comment](https://github.com/safe-global/safe-react-apps/pull/476#discussion_r923356145):

the flow now could be:

```
With an effect
1. We get an onChange event
2. We set the address in the state
3. We run an effect to detect if a proxy is present
4. If true we update the `implementationProxyAddress` variable state
```

and then in the JSX we can do:

```
address={implementationProxyAddress || address}
```

Reply:
I don't think that by decoupling effects, we win something. Checking if an address is a proxy is closer to input validation than a side effect to me. Having a single effect run on `onChange` seems more logical. Having separate asynchronous effects which are separate but ultimately depend on one trigger (user input) create unnecessary complexity, makes it harder to work with, and leads to sneaky bugs (one example would be the `AddressInput` component)

Reply:
I have some concerns about the amount of code in this change handler because in the future we probably will need to make more changes over as the address input is the starting point for the tx-builder flows.

I understand your concerns with effects but probably extracting this to a hook with the corresponding effects as @DaniSomoza suggested in the first comment will make it more maintainable .

Reply:
> I have some concerns about the amount of code in this change handler because in the future we probably will need to make more changes over as the address input is the starting point for the tx-builder flows.

I agree that this is a somewhat complex handler, but the logic is relatively straightforward and shouldn't be a problem to adjust. Not more complicated than trying to glue together four different effects to run a single algorithm.  

> I understand your concerns with effects but probably extracting this to a hook with the corresponding effects as @DaniSomoza suggested in the first comment will make it more maintainable .

I don't see it this way. In my eyes, it will not make it more maintainable but overcomplicate things and lead to unexpected bugs. I already mentioned an example of `AddressInput` where the same approach of splitting things into separate effects made the effects low-key dependent on one another and led to sneaky bugs.

Ultimately, what we want to do here, is to handle the user's change event. So, as a part of the handler, we run a few checks of the user input, and then we decide if we want to use this input or derive another value from the user's input. 

If we write a "pseudo-code", my solution is closer to this:
```
const [address, setAddress] = useState("")

const onChange = (e) => {
  const value = e.target.value

  const implementation = isProxy(value)
  if (implementation) {
    const implementationAbi = getImplementationAbi(implementation)
    if (implementationAbi) {
      setAddress(implementation)
    }
  } else {
    setAddress(value)
  }
}
```
The approach with effects looks like this:
```
const [address, setAddress] = useState("")
const [implementationAddress, setImplementationAddress] = useState("")

useEffect(() => {
  const implementation = isProxy(address)

  if (implementation) {
    setImplementationAddress(implementation)
  }
}, [address])

useEffect(() => {
  if (implementationAddress) {
    setAddress(implementationAddress)
  }

}, [implementationAddress])

const onChange = (e) => {
  setAddress(e.target.value)
}
```

I honestly do not see how the second approach is more maintainable or easier to understand. 
I can also quote some parts from the [react.js docs](https://beta.reactjs.org/learn/you-might-not-need-an-effect):

> You don’t need Effects to handle user events. For example, let’s say you want to send an /api/buy POST request and show a notification when the user buys a product. In the Buy button click event handler, you know exactly what happened. By the time an Effect runs, you don’t know what the user did (for example, which button was clicked). This is why you’ll usually handle user events in the corresponding event handlers.

> When you choose whether to put some logic into an event handler or an Effect, the main question you need to answer is what kind of logic it is from the user’s perspective. **If this logic is caused by a particular interaction, keep it in the event handler.** If it’s caused by the user seeing the component on the screen, keep it in the Effect.

Also, dan_abramov [tweet](https://twitter.com/dan_abramov/status/1281671006659260416) about `useEffect`:
> In hindsight, my personal conviction has become that if some effect can’t safely over-fire (e.g. fire twice instead of once), there is a problem. Usually things that can’t over-fire are related to user actions ("buy", "send", "finish"). Actions start in event handlers. Use them!